onnxruntime/core/providers/vitisai/imp/ep_context_utils.cc

```diff
@@ -0,0 +1,594 @@
+  p_attr_2->set_s(kVitisAIExecutionProvider);
+  // Attr "onnx_model_filename".
+  auto p_attr_3 = ONNX_NAMESPACE::AttributeProto::Create();
+  p_attr_3->set_name(kONNXModelFileNameAttr);
+  // p_attr_3->set_type(onnx::AttributeProto_AttributeType_STRING);
+  p_attr_3->set_type(ONNX_NAMESPACE::AttributeProto::STRING);
+  p_attr_3->set_s(fs::path(graph_viewer.ModelPath().ToPathString()).filename().string());
+  // Attr "notes".
+  auto p_attr_4 = ONNX_NAMESPACE::AttributeProto::Create();
+  p_attr_4->set_name(kNotesAttr);
+  // p_attr_4->set_type(onnx::AttributeProto_AttributeType_STRING);
+  p_attr_4->set_type(ONNX_NAMESPACE::AttributeProto::STRING);
+  // FIXME: 2G-limit of ProtoBuf.
+  if (saving_orig_graph) {
+    p_attr_4->set_s(SerializeOrigialGraph(graph_viewer));
+    LOGS_DEFAULT(VERBOSE) << "Saved serialized graph to attr proto";
+  } else {
+    nlohmann::json j_obj;
+    j_obj["backend_cache_dir"] = backend_cache_dir;
+    j_obj["backend_cache_key"] = backend_cache_key;
+    p_attr_4->set_s(j_obj.dump(-1, ' ', false, nlohmann::json::error_handler_t::replace));
+    LOGS_DEFAULT(VERBOSE) << "Saved backend cache key to attr proto";
+  }
+
+  auto p_node_attrs = NodeAttributes::Create();
+  constexpr int num_attrs = 5;
+  p_node_attrs->reserve(num_attrs);
+  p_node_attrs->emplace(kEmbedModeAttr, *p_attr_0);
+  p_node_attrs->emplace(kEPCacheContextAttr, *p_attr_1);
+  p_node_attrs->emplace(kSourceAttr, *p_attr_2);
+  p_node_attrs->emplace(kONNXModelFileNameAttr, *p_attr_3);
+  p_node_attrs->emplace(kNotesAttr, *p_attr_4);
+
+  ep_ctx_graph.AddNode(kEPContextOp, kEPContextOp, "", input_node_arg_ptrs, output_node_arg_ptrs, p_node_attrs.get(), kEPContextOpDomain);
+  auto res_status = ep_ctx_graph.Resolve();
+  ORT_ENFORCE(res_status.IsOK(), res_status.ErrorMessage());
+  LOGS_DEFAULT(VERBOSE) << "Created EP context model graph resolved";
+  if (ValidateEPContextNode(ep_ctx_graph)) {
+    LOGS_DEFAULT(VERBOSE) << "Created EP context model graph validated";
+  } else {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Created EP context model is invalid");
+  }
+
+  return Status::OK();
+}
+
+void DumpEPContextModel(
+    const std::unique_ptr<Model>& p_model, const std::string& ep_ctx_model_file_loc) {
+  std::fstream dump_stream(ep_ctx_model_file_loc, std::ios::out | std::ios::trunc | std::ios::binary);
+  p_model->ToProto()->SerializeToOstream(dump_stream);
+  LOGS_DEFAULT(VERBOSE) << "[VitisAI EP] Dumped " << ep_ctx_model_file_loc;
+}
+
+bool ValidateEPContextNode(const Graph& graph) {
+  // TODO: Support for multi-node EP context model.
+  LOGS_DEFAULT(VERBOSE) << "Number of nodes of EP context model: " << graph.Nodes().size();
+  assert(graph.Nodes().size() == 1);
+  auto* p_node = graph.GetNode(0);
+  assert(p_node->OpType() == kEPContextOp);
+  auto& attrs = p_node->GetAttributes();
+  assert(attrs.count(kEmbedModeAttr) > 0);
+  assert(attrs.count(kEPCacheContextAttr) > 0);
+  assert(attrs.count(kSourceAttr) > 0);
```

Comment:
What will happen if the EPContext node is generated/for other EP?
For QNN EP, we use EP name as the source if it's generated by the EP, or use "QNN" as source if it's generated by user script. We check for these key word in the EP code to make sure it's the EPContext node for the EP not for other EPs.

Reply:
1) Let me add one more explicit check in the function `ValidateEPContextNode()` for the content of of the attr `source`.
2) Why we can do (1) above is because we have set the attr `source` to `kVitisAIExecutionProvider` when an EP context node is created in the function `CreateEPContextModel()`.

Reply:
BTW -
1) Seems like TensorRT EP doesn't use the "source" attr at all.
2) QNN EP and OpenVINO EP, in the ONNXRT C++ implementation, only set "source" as the EP name accordingly.

Do you mind elaborating a bit on how user script set "qnn" for QNN EP?

Reply:
Here's the for QNN EP, https://github.com/microsoft/onnxruntime/blob/dadd0c451a60a6f908ef40fc93e7775765280c09/onnxruntime/python/tools/qnn/gen_qnn_ctx_onnx_model.py#L173-L183
It is used in our sample code:
https://github.com/microsoft/onnxruntime-inference-examples/blob/0de2e66e03981714e5308c457b72d785e98d0fe2/c_cxx/QNN_EP/mobilenetv2_classification/run_qnn_ep_sample.bat#L82

Reply:
Just added a source value "vitisai" for offline user scripts.